core/src/main/java/org/sonarsource/sonarlint/core/log/MessageFormat.java

```diff
@@ -0,0 +1,490 @@
+  // 16 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void handleLiteralChar(final StringBuilder buffer, final int escapeCounter, final char curChar) {
+    // any other char beside ESCAPE or DELIM_START/STOP-combo
+    // write unescaped escape chars
+    writeUnescapedEscapeChars(escapeCounter, buffer);
+    buffer.append(curChar);
+  }
+
+  /**
+   * Writes "{}" to the specified result array at the specified position and returns the resulting position.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 18 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void writeDelimPair(final StringBuilder buffer) {
+    buffer.append(DELIM_START);
+    buffer.append(DELIM_STOP);
+  }
+
+  /**
+   * Returns {@code true} if the specified parameter is odd.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 11 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static boolean isOdd(final int number) {
+    return (number & 1) == 1;
+  }
+
+  /**
+   * Writes a '\' char to the specified result array (starting at the specified position) for each <em>pair</em> of
+   * '\' escape chars encountered in the message format and returns the resulting position.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 11 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void writeEscapedEscapeChars(final int escapeCounter, final StringBuilder buffer) {
+    final int escapedEscapes = escapeCounter >> 1; // divide by two
+    writeUnescapedEscapeChars(escapedEscapes, buffer);
+  }
+
+  /**
+   * Writes the specified number of '\' chars to the specified result array (starting at the specified position) and
+   * returns the resulting position.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 20 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void writeUnescapedEscapeChars(int escapeCounter, final StringBuilder buffer) {
+    while (escapeCounter > 0) {
+      buffer.append(ESCAPE_CHAR);
+      escapeCounter--;
+    }
+  }
+
+  /**
+   * Appends the argument at the specified argument index (or, if no such argument exists, the "{}" delimiter pair) to
+   * the specified result char array at the specified position and returns the resulting position.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 25 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void writeArgOrDelimPair(final Object[] arguments, final int argCount, final int currentArgument,
+    final StringBuilder buffer) {
+    if (currentArgument < argCount) {
+      recursiveDeepToString(arguments[currentArgument], buffer, null);
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'recursiveDeepToString' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=165&issues=AWM_D9R25HoPdzsInowZ&open=AWM_D9R25HoPdzsInowZ)

[ISSUE_KEY]: AWM_D9R25HoPdzsInowZ